### PR TITLE
Permission name change on service provider authorizer

### PR DESCRIPTION
**Pull request by zahnno**
Noticed a use conflict with the _**serviceprovider.action.id**_ - we have two uses for these prefixed permissions. 

One is the ServiceProviderAwareDAO which adds a layer of filtering on DAO services allowing updating objects based on if the user has a variation of permission mentioned above (id replaced by service provider id). The other permits users to update read, remove service provider objects which is applied on the serviceProviderDAO authorizer. 

It'd be a good idea to make a distinction on these so I've made a change to the service provider dao's authorizer string that constructs the permission.

### CODE REVIEW

**Comment by noodlemoodle** [comment]
are non-admin users supposed to have read access to ServiceProviderDAO for their own spid? if so, you might need to add `serviceproviderdao.read.<spid>` to `ServiceProvider.inherentPermissions` as well. if not then lgtm

**Comment by lchanmann** [comment]
> the ServiceProviderAwareDAO which adds a layer of filtering on DAO services allowing updating objects based on spid if the user has the permission mentioned above

The "serviceprovider.update.id" check is only triggered when the user is trying to change/re-assign a different spid to the object. Otherwise, it check for "serviceprovider.read.id".

**Comment by zahnno** [comment]
@noodlemoodle I'm fine with that, any ideas when a user needs the service provider object now? The service provider object doesn't hold information a regular user would care to know afaik.

**Comment by zahnno** [comment]
@lchanmann ah yes, I should have made that clear since I referred to the update permission in my comment. Updated the PR summary.

**Comment by lchanmann** [comment]
IMO, "serviceproviderdao" permissions might not end up being used. The implication and impact of updating or removing a service provider at runtime hasn't been evaluated, we should not give "serviceproviderdao.update.id" or "serviceproviderdao.remove.id" permissions to anybody. Also not sure who might need "serviceproviderdao.read.id" permission. FYI @zahnno @jlhughes

**Comment by noodlemoodle** [comment]
they shouldn't need it, but we do have places were we call findspid for serviceprovideraware objects which calls the serviceproviderdao, which would probably cause side effects if the context user is non admin

**Comment by lchanmann** [comment]
for non-admin users the `ServiceProvider.inherentPermissions` should already give them access to find owned spid.

**Comment by noodlemoodle** [comment]
> for non-admin users the `ServiceProvider.inherentPermissions` should already give them access to find owned spid.

inherentpermission only gives serviceprovider.read, with this we woul need serviceproviderdao.read to find it no?

**Comment by zahnno** [comment]
> IMO, "serviceproviderdao" permissions might not end up being used.

I think this is irrelevant. Applying a non conflicting authorization layer (it already has an authorizer applied to it) is what is being resolved here. An authorized layer on a service is better than one that doesn't.

> The implication and impact of updating or removing a service provider at runtime hasn't been evaluated

I don't understand how runtime is a concern here. Your concern would be relevant to all authorizers that are handled by permissions and applied to services.

> Also not sure who might need "serviceproviderdao.read.id" permission.

As Ruby mentioned the system fetches spid using the user context in various areas - this could be fixed by replacing it with system context.

Regardless, spids are currently being read by users - the previous consideration was that it did not matter if a user can see its spid object.

By looking at the ServiceProvider - I don't see any reason why spid objects would be read sensitive. 

@noodlemoodle @lchanmann @jlhughes

**Comment by lchanmann** [comment]
So you wanted to differentiate between say "serviceprovider.remove.id" the permission to delete the service provider aware object eg. User as opposed to "serviceproviderdao.delete.id" the permission to delete the service provider the service provider itself?

If the user already have permission to delete object eg. "user.remove.id" why do they need "serviceprovider.remove.id" permission? and if someone has "serviceproviderdao.delete.id" permission then delete the service provider eg. "intuit", what should happen?

I am still hesitant on having "serviceproviderdao.action.id" permissions.

File: src/foam/nanos/auth/ServiceProvider.js [hunk 1]
```diff
@@ -57,7 +57,10 @@ foam.CLASS({
       name: 'inherentPermissions',
       javaGetter: 'return new String[] { "serviceprovider.read." + getId() };',
       factory: function() {
-        return [ 'serviceprovider.read.' + this.id ];
+        return [
+          'serviceprovider.read.' + this.id,
+          'serviceproviderdao.read.' + this.id
```
Comment (jlhughes) [inline src/foam/nanos/auth/ServiceProvider.js:65]: Why is this required, and this is only for javascript/client, as the inherit implements a javaGetter which supersedes this.

Comment (zahnno) [reply src/foam/nanos/auth/ServiceProvider.js:65]: User.findSpid(x) is called in various places and yes, I was just about to add it into the javagetter.

Comment (lchanmann) [reply src/foam/nanos/auth/ServiceProvider.js:65]: So it seems likely that when giving someone "serviceprovider.read" permission we should also give "serviceproviderdao.read" to the same user.

**Comment by lchanmann** [comment]
Rule.authorizeOnDelete() still uses "serviceprovider.update" permission, should it be updated?

Ref: https://github.com/kgrgreer/foam3/blob/6cca957a1fe875359cb83d32a6c25968b8f36104/src/foam/nanos/ruler/Rule.js#L481-L491

**Review by jlhughes — APPROVED** [review]

**Review by noodlemoodle — APPROVED** [review]

**Review by lchanmann — APPROVED** [review]
